File: demo/src/pages/ObserverableObjectDemo/ObserverableObjectDemo.swift
import SwiftUI
//https://www.hackingwithswift.com/quick-start/swiftui/how-to-use-observedobject-to-manage-state-from-external-objects
class UserProgress: ObservableObject {
    @Published var score = 0
}


struct InnerView: View {
    @ObservedObject var progress: UserProgress

    var body: some View {
        Button("Increase Score") {
            progress.score += 1
        }
    }
}

struct ObservableObjectDemo: View {
    @StateObject var progress = UserProgress()

    var body: some View {
        VStack {
            Text("Your score is \(progress.score)")
            InnerView(progress: progress)
        }
    }
}

File: demo/src/pages/ObserverableObjectDemo/ObserverableObjectDemo.ts
import {
    main, Published, Button, VStack, Text, ObservedObject, ObservableObject,
    Viewable, swifty, StateObject, Bound
} from "@tswift/ui";

class UserProgress extends ObservableObject {
    @Published score = 0
}
class InnerView extends Viewable {

    @ObservedObject progress: UserProgress;

    constructor(progress: UserProgress) {
        super();
        this.progress = progress;
    }

    body = Button("Increase Score", () => {
        this.progress.score += 1
    });

}



@main
export class ObserverableObjectDemo extends Viewable {
    @StateObject progress = new UserProgress();

    body = ({ progress }: Bound<this>) => VStack(
        Text(`Your score is ${progress.score}`),
        new InnerView(progress)
    );

}
